tests/failpoints/cases/test_ttl.rs

```diff
@@ -34,72 +37,110 @@ fn test_ttl_checker_impl<F: KvFormat>() {
         .path(dir.path())
         .api_version(F::TAG);
     let engine = builder.build_with_cfg(&cfg).unwrap();
-
     let kvdb = engine.get_rocksdb();
-    let key1 = b"zr\0key1";
+
+    // Make all entries earlier than safe point.
+    // TTL expired entries can only be collected when commit_ts < safe_point.
+    let commit_ts = 100;
+    let mut gc_runner = TestGcRunner::new(200);
+
+    let mut do_compact = |start_key: &[u8], end_key: &[u8], mut expect_keys: usize| {
+        gc_runner.prepare_gc(&kvdb);
+        check_ttl_and_compact_files(&kvdb, start_key, end_key, false);
+
+        if F::TAG == kvrpcpb::ApiVersion::V2 {
+            while let Ok(Some(task)) = gc_runner.gc_receiver.recv_timeout(Duration::from_secs(3)) {
+                match task {
+                    GcTask::RawGcKeys { keys, .. } => {
+                        expect_keys = expect_keys.checked_sub(keys.len()).unwrap();
+
+                        // Delete keys by `delete_cf` for simplicity.
+                        // In real cases, all old MVCC versions of `key` should be deleted.
+                        // See `GcRunner::raw_gc_keys`.
+                        for key in keys {
+                            let db_key =
+                                keys::data_key(key.append_ts(commit_ts.into()).as_encoded());
+                            kvdb.delete_cf(CF_DEFAULT, &db_key).unwrap();
+                        }
+
+                        if expect_keys == 0 {
+                            break;
+                        }
+                    }
+                    _ => unreachable!(),
+                }
+            }
+            kvdb.flush_cf(CF_DEFAULT, true).unwrap();
+            assert_eq!(expect_keys, 0);
+        }
+
+        gc_runner.post_gc();
+    };
+
+    let key1 = make_raw_key::<F>(b"r\0key1", Some(commit_ts));
```

Comment:
How about encapsulating `make_raw_key`, make RawValue, `put_cf` and `flush_cf `  into a function?

Reply:
I refactor to a loop to generate test data, for eliminating duplicated codes. PTAL~
https://github.com/tikv/tikv/pull/15143/commits/bfa18d88923ff4ab822003abe16ca618c73a4c64